### PR TITLE
Implement digamma

### PR DESCRIPTION
**Pull request by zou3519**
Implements `torch.digamma` and `tensor.digamma()` as per #678

cc @fritzo 

### Test Plan
New unit tests. cpu unit tests test some edge cases, gpu unit test compares output of cpu to gpu

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -194,6 +194,24 @@ def test_sinh(self):
     def test_lgamma(self):
         self._testMathByName('lgamma')
 
+    def test_digamma(self):
+        y = torch.Tensor([-10, 0])
+        x = torch.Tensor([-0.1, 3, 999])
```
Comment (fritzo) [inline test/test_torch.py]: You could also check against hard-coded values generated by `scipy.special.digamma()`, e.g.:
```py
>>> import numpy as np
>>> import scipy.special
>>> x = 10.0 ** np.arange(-5, 5, 0.5)
>>> x
array([  1.00000000e-05,   3.16227766e-05,   1.00000000e-04,
         3.16227766e-04,   1.00000000e-03,   3.16227766e-03,
         1.00000000e-02,   3.16227766e-02,   1.00000000e-01,
         3.16227766e-01,   1.00000000e+00,   3.16227766e+00,
         1.00000000e+01,   3.16227766e+01,   1.00000000e+02,
         3.16227766e+02,   1.00000000e+03,   3.16227766e+03,
         1.00000000e+04,   3.16227766e+04])
>>> scipy.special.digamma(x)
array([ -1.00000577e+05,  -3.16233538e+04,  -1.00005771e+04,
        -3.16285436e+03,  -1.00057557e+03,  -3.16799792e+02,
        -1.00560885e+02,  -3.21491437e+01,  -1.04237549e+01,
        -3.31321826e+00,  -5.77215665e-01,   9.84925052e-01,
         2.25175259e+00,   3.43798293e+00,   4.60016185e+00,
         5.75488076e+00,   6.90725520e+00,   8.05888970e+00,
         9.21029037e+00,   1.03616171e+01])
```

**Review by fritzo — COMMENTED** [review]
Thanks for implementing this!

**Comment by gchanan** [comment]
fwiw I'm adding the ability to write efficient pointwise native functions in ATen directly.

**Comment by fritzo** [comment]
Does this PR also register `digamma` as the derivative of `lgamma`? (I'm not sure how that works.)

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -305,7 +305,10 @@
   end: grad * weight
 
 - name: lgamma(Tensor self)
-  self: not_implemented("lgamma")
+  self: grad * digamma(self)
```
Comment (fritzo) [inline tools/autograd/derivatives.yaml]: Do I understand correctly that this is all that is needed to register `digamma` as the grad of `lgamma`?

File: aten/src/TH/generic/THTensorMath.c [hunk 3]
```diff
@@ -3092,6 +3092,81 @@ void THTensor_(atan2)(THTensor *r_, THTensor *tx, THTensor *ty)
...
+
+  int x_is_integer = x == floor(x);
+  if (x < 0) {
+    if (x_is_integer) {
+      return INFINITY;
+    }
+    return THTensor_(digamma_one)(1 - x) - PI / tan(PI * x);
+  }
+
+	// Push x to be >= 10
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: nit: tabs

File: aten/src/TH/generic/THTensorMath.c [hunk 4]
```diff
@@ -3092,6 +3092,81 @@ void THTensor_(atan2)(THTensor *r_, THTensor *tx, THTensor *ty)
...
+  accreal y = 0;
+  if (x < 1.0e17) {
+	  accreal z = 1.0 / (x * x);
+	  y = z * THTensor_(polevl)(z, A, 6);
+  }
+  return log(x) - (0.5 / x) - y;
+}
+
+/*
+ * Algorithm adapted from Cephes
```
Comment (apaszke) [inline aten/src/TH/generic/THTensorMath.c]: Does adapted == pretty much copied, or did you rewrite it based on Cephes? I'm not sure what's the license, but we might need to include it here.

File: test/test_cuda.py [hunk 5]
```diff
@@ -1032,6 +1033,23 @@ def test_var_stability(self):
         tensor = tensor.unsqueeze(1)
         self.assertEqual(tensor.var(0)[0], 0.03125)
 
+    def test_digamma(self):
```
Comment (apaszke) [inline test/test_cuda.py]: Any reason why `digamma` can't be tested like all other functions?

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (apaszke) [reply tools/autograd/derivatives.yaml]: Yes, that's correct

**Review by apaszke — COMMENTED** [review]
Looks good. We should resolve the licensing issues before merging this.

File: aten/src/TH/generic/THTensorMath.c (hunk 4 above)
Comment (zou3519) [reply aten/src/TH/generic/THTensorMath.c]: I think the answer to that is pretty much copied: the algorithm is the same, except for one case (integers from 1 - 10). Here's the original cephes code I used for reference: https://github.com/jeremybarnes/cephes/blob/master/misc/psi.c -- I can include the license there here.

File: test/test_cuda.py (hunk 5 above)
Comment (zou3519) [reply test/test_cuda.py]: The values of digamma on gpu when compared to cpu are the same to 6 digits. I tried making this a normal test and some output values were high (~100), resulting in an error of around 1e-3, which is above the error threshold

File: aten/src/TH/generic/THTensorMath.c (hunk 4 above)
Comment (apaszke) [reply aten/src/TH/generic/THTensorMath.c]: Cephes doesn't state its license clearly except from saying that it "can be used freely". I think it's best to ask the author, just to be sure. Aparently [that's what authors of torch-cephes did](https://github.com/deepmind/torch-cephes/blob/master/LICENSE.txt).

**Comment by fritzo** [comment]
FYI this will merge conflict with #3978 where I've implemented a makeshift finite-difference version of `THTensor(digamma_one)`. If #3978 merges before this PR, you can simply delete my makeshift version.

**Comment by fritzo** [comment]
@zou3519 can I help with this at all? Now that you've done the hard part, I'm happy to resolve merge conflicts and add tests. (We're looking forward to using this in Pyro, and we already have a wrapper to use `torch.distributions.Gamma`).

**Comment by zou3519** [comment]
@fritzo I haven't reached out to the author about the licensing yet. I'll do that tomorrow and we'll see how that goes :)

**Comment by fritzo** [comment]
I see, thanks for letting me know!

**Comment by ezyang** [comment]
Any update on the license status?

**Comment by zou3519** [comment]
Sent an email a few days ago to the author, haven't heard back yet

**Comment by fritzo** [comment]
If we still haven't heard from the author, I could throw together a little PR that exposes the makeshift finite-difference implementation of digamma that we're already using internally in a few places. This would at least unblock users of Gamma, Beta, and Dirichlet distributions.
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/aten/src/ATen/native/Distributions.cpp#L21
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/aten/src/TH/generic/THTensorMath.c#L3483
- https://github.com/pytorch/pytorch/blob/15b657af843d0dd95fa43d44348161d3785ed0e9/torch/distributions/utils.py#L6

**Comment by apaszke** [comment]
I think merging an approximation is good for now

**Comment by fritzo** [comment]
Hmm I looked into it but it appears to be quite complex now that our functions using `digamma` are spread across C code in `aten/src/TH/generic/THTensorMath.c` and C++ code in `aten/src/ATen/native/Distributions.cpp`. @gchanan would it be difficult for you to move `dirichlet_grad()` to C++ as well so we can implement `digamma()` in a single place?

**Comment by gchanan** [comment]
@fritzo sure, I'll take a look at.

**Comment by zou3519** [comment]
Licensing issues resolved, PR rebased. I was told that "You are welcome to modify and distribute the library under BSD license." so I've added the original copyright notices into the code as comments.

**Comment by fritzo** [comment]
That's great news, @zou3519 !

**Comment by wranai** [comment]
Any news on this? It seems only the GPU build failed on Windows last November and that only because it run out of memory, but everything else was okay.

**Comment by fritzo** [comment]
@wranai A simpler implementation of digamma and trigamma is already in master. Try `torch.digamma`

**Pull request closed without merging**